spec.md

```diff
@@ -1090,6 +1119,287 @@ message ControllerServiceCapability {
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Snapshot already exists but is incompatible | 6 ALREADY_EXISTS | Indicates that a snapshot corresponding to the specified snapshot `name` already exists but is incompatible with the specified `volume_id`. | Caller MUST fix the arguments or use a different `name` before retrying. |
+| Operation pending for snapshot | 10 ABORTED | Indicates that there is a already an operation pending for the specified snapshot. In general the Cluster Orchestrator (CO) is responsible for ensuring that there is no more than one call "in-flight" per snapshot at a given time. However, in some circumstances, the CO MAY lose state (for example when the CO crashes and restarts), and MAY issue multiple calls simultaneously for the same snapshot. The Plugin, SHOULD handle this as gracefully as possible, and MAY return this error code to reject secondary calls. | Caller SHOULD ensure that there are no other calls pending for the specified snapshot, and then retry with exponential back off. |
+| Call not implemented | 12 UNIMPLEMENTED | CreateSnapshot call is not implemented by the plugin or disabled in the Plugin's current mode of operation. | Caller MUST NOT retry. Caller MAY call `ControllerGetCapabilities` to discover Plugin capabilities. |
+
+##### CreateSnapshot vs GetSnapshot
+
+In addition to `CreateSnapshot`, a `GetSnapshot` function is also introduced.
+
+For some cloud providers, a snapshot will be uploaded to a location in the cloud (i.e., an object store) after the snapshot is cut (created).
+For other cloud providers or storage backends, a snapshot will be cut (created) but won’t be uploaded.
+The time it takes to upload a snapshot varies, ranging from a few minutes to a few hours, depending on how much data is stored on the snapshot.
+For this reason two interfaces are needed, one is `CreateSnapshot` and the other one is `GetSnapshot`.
+`CreateSnapshot` CSI function must block until the snapshot is cut.
+It should not wait for the snapshot to be uploaded.
+`GetSnapshot` CSI function should block until the upload is complete and the snapshot is ready.
+`GetSnapshot` should be called right after `CreateSnapshot`.
+
+`GetSnapshot` is not required. However if the `REQUIRES_GET_SNAPSHOT` capability is supported, CO must call `GetSnapshot` after calling `CreateSnapshot`.
+
+From CSI plugin’s point of view, both calls are synchronous.
+All CSI functions have been designed to be synchronous, so the snapshot functions are designed to follow the convention.
+It is important to separate these two functions so that a post-action such as `thaw` can be performed on the application or filesystem right after `CreateSnapshot` is complete.
+Since the uploading process could take hours, we don’t want to wait for the snapshot to be uploaded before performing `thaw` because the application should be made available as soon as possible.
+
+#### `GetSnapshot`
```

Comment:
I don't see a need for this (GetSnapshot) RPC. ListSnapshots provides the same (and more) functionality. If there's concern about iterating to pages of snapshot objects just to get details for a single snapshot, then perhaps amend the ListSnapshots RPC with some kind of filter-by-snapshot-id?